pkg/secrets/pushtofile/secrets_groups.go

```diff
@@ -0,0 +1,220 @@
+)
+
+type SecretsFileFormat int
+
+const (
+	FILE_FORMAT_NONE SecretsFileFormat = iota
+	FILE_FORMAT_YAML
+	FILE_FORMAT_JSON
+	FILE_FORMAT_DOTENV
+	FILE_FORMAT_BASH
+	FILE_FORMAT_PLAINTEXT
+)
+
+// SecretsPaths comprises Conjur variable paths for all secrets in a secrets
+// group, indexed by secret name.
+type SecretsPaths map[string]string
+
+// SecretsGroupInfo comprises secrets mapping information for a given secrets
+// group.
+type SecretsGroupInfo struct {
+	Secrets      SecretsPaths
+	FilePath     string
+	FileFormat   SecretsFileFormat
+	FilePerms    os.FileMode
+	FileTemplate string
+}
+
+// SecretsGroups comprises secrets mapping info for all secret groups
+type SecretsGroups map[string]SecretsGroupInfo
+
+func ExtractSecretsGroupsFromAnnotations(annotations map[string]string) (SecretsGroups, error) {
+	secretsGroups := make(SecretsGroups)
+
+	for annotation := range annotations {
+		if strings.HasPrefix(annotation, CONJUR_SECRETS_PREFIX) {
+			groupName := strings.TrimPrefix(annotation, CONJUR_SECRETS_PREFIX)
+
+			secretsPathPrefix, err := parseConjurSecretsPathPrefix(groupName, annotations[CONJUR_SECRETS_POLICY_PATH_PREFIX+groupName])
+			if err != nil {
+				return nil, err
+			}
+
+			secrets, err := parseConjurSecretsPaths(groupName, annotations[CONJUR_SECRETS_PREFIX+groupName], secretsPathPrefix)
+			if err != nil {
+				return nil, err
+			}
+
+			filePath, err := parseFilePath(groupName, annotations[SECRET_FILE_PATH_PREFIX+groupName])
+			if err != nil {
+				return nil, err
+			}
+
+			fileFormat, err := parseFileFormat(groupName, annotations[SECRET_FILE_FORMAT_PREFIX+groupName])
+			if err != nil {
+				return nil, err
+			}
+
+			fileTemplate := annotations[SECRET_FILE_TEMPLATE_PREFIX+groupName]
+
+			if fileTemplate != "" {
+				fileFormat = FILE_FORMAT_NONE
+			}
+
+			groupInfo := SecretsGroupInfo{
+				Secrets:      secrets,
+				FilePath:     filePath,
+				FileFormat:   fileFormat,
+				FilePerms:    DEFAULT_FILE_PERMS,
+				FileTemplate: fileTemplate,
+			}
+
+			err = validateGroupInfo(groupName, groupInfo)
+			if err != nil {
+				return nil, err
+			}
+
+			secretsGroups[groupName] = groupInfo
+		}
+	}
+
+	return secretsGroups, nil
+}
+
+func parseConjurSecretsPaths(groupName string, secretsPaths string, secretsPathsPrefix string) (SecretsPaths, error) {
+	secrets := make(SecretsPaths)
+
+	unpacked := []interface{}{}
+	if err := yaml.Unmarshal([]byte(secretsPaths), &unpacked); err != nil {
+		return nil, log.RecordedError(messages.CSPFK050E, err.Error(), groupName)
+	}
+
+	insertSecret := func(name string, secretPath string) error {
+		secretPath = path.Clean(secretPath)
+		if secretPath == "." {
+			return log.RecordedError(messages.CSPFK050E, "Invalid secret path", groupName)
+		}
+		secrets[name] = secretsPathsPrefix + secretPath
+
+		return nil
+	}
+
+	for _, secret := range unpacked {
+		switch val := secret.(type) {
+		case string:
+			{
+				name := path.Base(val)
```

Comment:
This is another case where we use the path package which assumes file paths, and we get some funky behaviour because policy paths are not file paths. Beware edge cases! For example when `val` is an empty string then name becomes `.` which is not what we want.

We could roll our own with:
```
name := val[strings.LastIndex(val, "/")+1:]
```